Modules/_json.c

```diff
@@ -1502,10 +1503,15 @@ encoder_encode_key_value(PyEncoderObject *s, _PyUnicodeWriter *writer, bool *fir
         return 0;
     }
     else {
-        PyErr_Format(PyExc_TypeError,
-                     "keys must be str, int, float, bool or None, "
-                     "not %.100s", Py_TYPE(key)->tp_name);
-        return -1;
+        newobj = PyObject_CallOneArg(s->defaultfn, key);
+        if (newobj == NULL) {
+            PyErr_Format(PyExc_TypeError,
+                         "keys must be str, int, float, bool or None, "
+                         "not %.100s", Py_TYPE(key)->tp_name);
+            return -1;
+        }
+        return encoder_encode_key_value(s, writer, first, newobj, value,
```

Comment:
Would it be necessary to decref newobj?

Reply:
Ah yes, I've now made the logics non-recursive to be consistent with the Python implemetation, and decref `newobj` after a successful key conversion.